app-engine/java/src/com/google/cloud/tools/intellij/appengine/java/sdk/ManagedCloudSdkService.java

```diff
@@ -367,21 +370,39 @@ public void onFailure(Throwable t) {
             .notifyManagedSdkJobFailure(jobType, t.toString());
       }
 
-      // failed or interrupted update might still keep SDK itself installed.
+      // update status of the SDK and generate corresponding metric events.
       switch (jobType) {
         case INSTALL:
           updateStatus(SdkStatus.NOT_AVAILABLE);
 
-          UsageTrackerProvider.getInstance()
-              .trackEvent(GctTracking.MANAGED_SDK_FAILED_INSTALL)
-              .ping();
+          FluentTrackingEventWithMetadata trackingEvent;
+          if (jobCancelled) {
+            trackingEvent =
+                UsageTrackerProvider.getInstance()
+                    .trackEvent(GctTracking.MANAGED_SDK_CANCELLED_INSTALL);
+          } else {
+            trackingEvent =
+                UsageTrackerProvider.getInstance()
+                    .trackEvent(GctTracking.MANAGED_SDK_FAILED_INSTALL)
+                    .addMetadata(GctTracking.METADATA_MESSAGE_KEY, t.toString());
+          }
+          trackingEvent.ping();
           break;
         case UPDATE:
+          // failed or interrupted update might still keep SDK itself installed.
           checkSdkStatusAfterFailedUpdate();
 
-          UsageTrackerProvider.getInstance()
-              .trackEvent(GctTracking.MANAGED_SDK_FAILED_UPDATE)
-              .ping();
+          if (jobCancelled) {
+            trackingEvent =
+                UsageTrackerProvider.getInstance()
+                    .trackEvent(GctTracking.MANAGED_SDK_CANCELLED_UPDATE);
+          } else {
+            trackingEvent =
```

Comment:
can you extract these pings into methods since they seem to be duplicated twice?